Improve Philips Hue color conversion 2

starkillerOG opened the pull request:
## Description:

See PR: https://github.com/home-assistant/home-assistant/pull/19123
This is just closing the previous PR and opening this one in order to do a rebase from within the web UI.
(Philips Hue files changed location, that is why the rebase was nessesary)


## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

starkillerOG committed:
Add gamut capability to color util

starkillerOG committed:
Include gamut in hue_test

starkillerOG committed:
Improve Philips Hue color conversion

starkillerOG committed:
correct import for new location hue.light

starkillerOG committed:
include file changes between PR's

starkillerOG committed:
update aiohue version

starkillerOG committed:
update aiohue version

starkillerOG committed:
update aiohue version

starkillerOG commented:
@balloob I think this PR is now ready to merge.
Are there any other suggestions or inprovements?

amelchio reviewed (changes requested):
Thank you for you persistence! I know that reworking your work several times is no fun.

I think this is looking good now, just missing some tests for the extended color util functionality.

starkillerOG commented:
@amelchio what kind of test would be needed?
I am not sure how to write tests, have never done that before.
I always just test locally and manually, but do not axactly know how to write test for Travis.

balloob commented:
We need to make sure that the things we have in our code keep working, even if aiohue gets upgrades etc. So `hue/test_light.py` needs to test different gamuts. That can either be added to existing tests if possible, or as a new test.

For the color functions, check `tests/util/test_color.py` and add a test for each new feature or function

starkillerOG committed:
fix hue_test

Now Idea why it failed compared to the previous time

starkillerOG committed:
Include gamut in hue_test

starkillerOG committed:
fix hue_test

starkillerOG committed:
Try to test hue gamut conversion

supply a color that is well outside the color gamut of the light, and see if the response is correctly converted to within the reach of the light.

starkillerOG committed:
switch from gamut A to gamut B for the tests.

starkillerOG committed:
remove white space in blanck line

starkillerOG committed:
Fix gamut hue test

starkillerOG committed:
Add Gamut tests for the util.color

starkillerOG committed:
fix hue gamut test

starkillerOG committed:
fix hue gamut test

starkillerOG committed:
Improve Philips Hue color conversion

starkillerOG commented:
@amelchio and @balloob I added a test to the `components/hue/test_light.py` to test if the gamut conversion is going alright in the turn_on function of the hue light.
I also added a hole bunch of tests to the  `tests/util/test_color.py` to check the conversion functions including a gamut. All those test have now succesfully passed.

Do you have suggestions for any more tests, or can this PR now be merged?

balloob commented:
Awesome !

starkillerOG commented:
@balloob and @amelchio Thank you for all the help and support during this PR!
I hope a lot of people can benefit from the new code.
I am glad that it is finally merged:)
